core/src/main/java/fr/sncf/osrd/standalone_sim/SignalProjection.kt

```diff
@@ -0,0 +1,262 @@
+    )
+    val signalUpdates = signalUpdates(pathSignals, signalAspectChangeEvents, loadedSignalInfra, rawInfra, signalSightings)
+    return SignalProjectionResult(signalUpdates)
+}
+
+// Returns all the signals on the path
+private fun pathSignals(
+    startOffset: Distance,
+    blockPath: StaticIdxList<Block>,
+    blockInfra: BlockInfra,
+    rawInfra: SimInfraAdapter
+): MutableList<PathSignal> {
+    val pathSignals = mutableListOf<PathSignal>()
+    var currentOffset = startOffset
+    for ((blockIdx, block) in blockPath.withIndex()) {
+        val blockSignals = blockInfra.getBlockSignals(block)
+        val blockSignalPositions = blockInfra.getSignalsPositions(block)
+        val numExclusiveSignalInBlock =
+            if (blockIdx == blockPath.size - 1) blockSignals.size else blockSignals.size - 1
+        for ((signal, position) in blockSignals.zip(blockSignalPositions).take(numExclusiveSignalInBlock)) {
+            pathSignals.add(PathSignal(signal, currentOffset + position))
+        }
+
+        for (zonePath in blockInfra.getBlockPath(block)) {
+            currentOffset += rawInfra.getZonePathLength(zonePath)
+        }
+    }
+
+    return pathSignals
+}
+
+
+private fun computeSignalAspectChangeEvents(
+    blockPath: StaticIdxList<Block>,
+    zoneToPathIndexMap: Map<String, Int>,
+    blockInfra: BlockInfra,
+    pathSignals: List<PathSignal>,
+    zoneUpdates: List<ZoneUpdate>,
+    simulator: SignalingSimulator,
+    rawInfra: SimInfraAdapter,
+    loadedSignalInfra: LoadedSignalInfra
+): Map<PathSignal, MutableList<SignalAspectChangeEvent>> {
+    val zoneCount = blockPath.sumOf { blockInfra.getBlockPath(it).size }
+    val zoneStates = ArrayList<ZoneStatus>(zoneCount)
+    for (i in 0 until zoneCount) zoneStates.add(ZoneStatus.CLEAR)
+
+    val signalAspects = pathSignals.associateBy({ it.signal }, { "VL" })
+        .toMutableMap() // TODO: Have a better way to get the least restrictive aspect
+
+    val signalAspectChangeEvents = pathSignals.associateBy({ it }, { mutableListOf<SignalAspectChangeEvent>() })
+    for (event in zoneUpdates) {
+        if (!zoneToPathIndexMap.containsKey(event.zone))
+            continue
+        if (event.isEntry)
+            zoneStates[zoneToPathIndexMap[event.zone]!!] = ZoneStatus.OCCUPIED
+        else
+            zoneStates[zoneToPathIndexMap[event.zone]!!] = ZoneStatus.CLEAR
+
+        val simulatedSignalStates = simulator.evaluate(
+            rawInfra, loadedSignalInfra, blockInfra, blockPath, 0, blockPath.size, zoneStates
+        )
+        val simulatedAspects = simulatedSignalStates.map { it.getEnum("aspect") }
+        for (pathSignal in pathSignals) {
+            val signal = pathSignal.signal
+            val aspect = simulatedAspects[signal] ?: continue
+            if (signalAspects[signal]!! == aspect) continue
+            signalAspectChangeEvents[pathSignal]!!.add(SignalAspectChangeEvent(aspect, (event.time * 1000).toLong(), event.offset.meters))
```

Comment:
It's generally difficult to follow which times are in seconds or milliseconds

Reply:
Ack and agreed We should have a time class similar to the distance one

Reply:
I'll open an issue for later, as discussed on slack

Reply:
https://github.com/DGEXSolutions/osrd/issues/3730